### PR TITLE
mining: Move block chain functions to Config.

### PR DESCRIPTION
**Pull request by rstaudt2**
This updates mining to take in the block chain related functions that it requires rather than requiring a blockchain.BlockChain instance.

This allows for alternate implementations to be provided for these functions, such as for testing.

Some of the functions, such as `BlockByHash(hash *chainhash.Hash)`, are only used for the tip and therefore _could_ have been defined as something like `BestBlock()` instead (which would internally call `BestSnapshot` and pass that hash to `BlockByHash`).  I decided against this since mining already has already fetched `blockchain.BestState` in most cases, so it would be a bit less efficient to be calling `BestSnapshot` again within these functions.

### CODE REVIEW

**Comment by davecgh** [comment]
> Some of the functions, such as `BlockByHash(hash *chainhash.Hash)`, are only used for the tip and therefore _could_ have been defined as something like `BestBlock()` instead (which would internally call `BestSnapshot` and pass that hash to `BlockByHash`). I decided against this since mining already has already fetched `blockchain.BestState` in most cases, so it would be a bit less efficient to be calling `BestSnapshot` again within these functions.

Regarding this point, one of the things I really want to work towards changing in the future is ensuring that all of the funcs end up taking in a block hash as opposed to assuming the current tip.  While it works with a high degree of accuracy due to the fact blocks are extremely infrequent relative to the execution time, the current logic is realistically racy in general due to the fact that the tip might have changed in between the point you call two funcs which rely on implicitly working with it since they independently lock and unlock.

I should also note that in all contexts which this occurs, it doesn't really have super negative consequences because it either produces a template the fails which will cause another one to be built which won't since the new attempt is super unlikely to also have the tip change in between due to target block times.  Other cases are similar, but it can very infrequently result in needless retries and even potentially slightly incorrect and/or outdated data delivered over RPC.

**Comment by rstaudt2** [comment]
> If all of the funcs end up taking the block hash, there would be a single call to determine the current tip hash and then that hash would be used in all of the calls which would prevent the aforementioned possibilities altogether.

👍 Makes sense, sounds like a good approach to me.

File: internal/mining/mining.go [hunk 1]
```diff
@@ -40,6 +40,134 @@ const (
...
+	CalcNextRequiredDifficulty func(hash *chainhash.Hash, timestamp time.Time) (uint32, error)
+
+	// CalcStakeVersionByHash defines the function to use to calculate the expected
+	// stake version for the block AFTER the provided block hash.
+	CalcStakeVersionByHash func(hash *chainhash.Hash) (uint32, error)
+
+	// CheckConnectBlockTemplate defines the function to use to fully validate that
+	// connecting the passed block to either the tip of the main chain or its
+	// parent does not violate any consensus rules, aside from the proof of work
+	// requirement.  The block must connect to the current tip of the main chain or
```
Comment (dnldd) [inline internal/mining/mining.go:91]: `The block must connect to the current tip of the main chain or its parent`, this is already stated in the sentence prior.

File: internal/mining/mining.go [hunk 2]
```diff
@@ -40,6 +40,134 @@ const (
...
+	TimeSource blockchain.MedianTimeSource
+
+	// SubsidyCache defines a subsidy cache to use.
+	SubsidyCache *standalone.SubsidyCache
+
+	// ChainParams defines the chain parameters to use.
+	ChainParams *chaincfg.Params
+
+	// BlockManager defines the block manager to use.
+	BlockManager blockManagerFacade
```
Comment (dnldd) [inline internal/mining/mining.go:63]: Considering the detail provided for the other config fields these could be more descriptive.

File: internal/mining/mining.go [hunk 3]
```diff
@@ -40,6 +40,134 @@ const (
...
+type Config struct {
+	// Policy houses the policy (configuration parameters) which is used to control
+	// the generation of block templates.
+	Policy *Policy
+
+	// TxSource represents a source of transactions to consider for inclusion in
+	// new blocks.
+	TxSource TxSource
+
+	// TimeSource defines a time source to use.
```
Comment (davecgh) [inline internal/mining/mining.go:53]: These comments should be made a bit more explicit like the others as opposed to stating the obvious.  They don't seem to be especially useful as they are.

**Review by davecgh — CHANGES_REQUESTED** [review]
Looks great overall and it's a good set of changes.

**Comment by rstaudt2** [comment]
Thanks for the reviews.  I adjusted the comments on items that were called out to be more verbose.

**Review by davecgh — APPROVED** [review]
Thanks for the updates.  Looks good!

**Review by dnldd — APPROVED** [review]

**Pull request merged**